File: exp-12/q2-floyd-warshall/solution.js
const fs = require("fs");

const data = fs.readFileSync(0, "utf8").trim().split(/\s+/).map(Number);
let idx = 0;
const n = data[idx++];

const INF = 999999;
const dist = [];
for (let i = 0; i < n; i++) {
  const row = [];
  for (let j = 0; j < n; j++) {
    row.push(data[idx++]);
  }
  dist.push(row);
}

for (let k = 0; k < n; k++) {
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      if (
        dist[i][k] !== INF &&
        dist[k][j] !== INF &&
        dist[i][k] + dist[k][j] < dist[i][j]
      ) {
        dist[i][j] = dist[i][k] + dist[k][j];
      }
    }
  }
}

for (let i = 0; i < n; i++) {
  const row = [];
  for (let j = 0; j < n; j++) {
    row.push(dist[i][j] === INF ? "INF" : dist[i][j]);
  }
  console.log(row.join(" "));
}

File: exp-12/q2-floyd-warshall/solution.py
import sys

INF = 999999


def main():
    data = sys.stdin.read().split()
    idx = 0
    n = int(data[idx]); idx += 1

    dist = []
    for _ in range(n):
        row = []
        for _ in range(n):
            row.append(int(data[idx])); idx += 1
        dist.append(row)

    for k in range(n):
        for i in range(n):
            for j in range(n):
                if dist[i][k] != INF and dist[k][j] != INF and \
                   dist[i][k] + dist[k][j] < dist[i][j]:
                    dist[i][j] = dist[i][k] + dist[k][j]

    for i in range(n):
        for j in range(n):
            if j > 0:
                print(" ", end="")
            if dist[i][j] == INF:
                print("INF", end="")
            else:
                print(dist[i][j], end="")
        print()


if __name__ == "__main__":
    main()
